File: yggdrasil/examples/wofost/src/wofost.f90
program main
  ! Include methods for input/output channels
  use fygg

  ! Declare resulting variables and create buffer for received message
  logical :: flag = .true.
  type(yggcomm) :: in_channel, out_channel
  type(ygggeneric) :: obj
  character(len=20), dimension(:), pointer :: keys
  type(ygggeneric), pointer :: amaxtb
  real(kind=8), dimension(:), pointer :: amaxtb_x, amaxtb_y
  real(kind=8), pointer :: co2
  integer :: i
  obj = init_generic()

  ! Initialize input/output channels
  in_channel = ygg_generic_input("input")
  out_channel = ygg_generic_output("output")

  ! Loop until there is no longer input or the queues are closed
  do while (flag)

     ! Receive input from input channel
     ! If there is an error, the flag will be false
     ! Otherwise, it is the number of variables filled
     flag = ygg_recv_var(in_channel, yggarg(obj))
     if (.not.flag) then
        write (*, '("Fortran Model: No more input.")')
        exit
     end if

     ! Print received message
     write (*, '("Fortran Model:")')
     call display_generic(obj)

     ! Print keys
     call generic_map_get_keys(obj, keys)
     write (*, '("Fortran Model: keys = ")')
     do i = 1, size(keys)
        write (*, '(A," ")', advance="no") trim(keys(i))
     end do
     write (*, '("")')

     ! Get double precision floating point element
     call generic_map_get(obj, "CO2", co2)
     write (*, '("Fortran Model: CO2 = ",F10.5)') co2

     ! Get array element
     call generic_map_get(obj, "AMAXTB", amaxtb)
     call generic_array_get(amaxtb, 1, amaxtb_x)
     call generic_array_get(amaxtb, 2, amaxtb_y)
     write (*, '("Fortran Model: AMAXTB = ")')
     do i = 1, size(amaxtb_x)
        write (*, '(A,F10.5,A,F10.5)') char(11), amaxtb_x(i), &
             char(11), amaxtb_y(i)
     end do

     ! Send output to output channel
     ! If there is an error, the flag will be false
     flag = ygg_send_var(out_channel, yggarg(obj))
     if (.not.flag) then
        write (*, '("Fortran Model: Error sending output.")')
        exit
     end if

  end do

end program main

File: yggdrasil/examples/wofost/src/wofost.cpp
#include <iostream>
// Include methods for input/output channels
#include "YggInterface.hpp"

int main(int argc, char *argv[]) {
  // Initialize input/output channels
  YggJSONObjectInput in_channel("input");
  YggJSONObjectOutput out_channel("output");

  // Declare resulting variables and create buffer for received message
  int flag = 1;
  rapidjson::Document obj;
  rapidjson::CrtAllocator allocator;
  double* amaxtb_x = NULL;
  double* amaxtb_y = NULL;
  char** keys = NULL;
  size_t nkeys, i;
  double co2;
  rapidjson::SizeType n_amaxtb;

  // Loop until there is no longer input or the queues are closed
  while (flag >= 0) {
  
    // Receive input from input channel
    // If there is an error, the flag will be negative
    // Otherwise, it is the size of the received message
    flag = in_channel.recv(1, &obj);
    if (flag < 0) {
      std::cout << "C++ Model: No more input." << std::endl;
      break;
    }

    // Print received message
    std::cout << "C++ Model:" << std::endl <<
      document2string(obj) << std::endl;

    // Get double precision floating point element
    co2 = obj["CO2"].GetDouble();
    std::cout << "C++ Model: CO2 = " << co2 << std::endl;

    // Get array element
    const rapidjson::Value& amaxtb = obj["AMAXTB"];
    amaxtb[0].Get1DArray(amaxtb_x, n_amaxtb, obj.GetAllocator());
    amaxtb[1].Get1DArray(amaxtb_y, n_amaxtb, obj.GetAllocator());
    std::cout << "C++ Model: AMAXTB = " << std::endl;
    for (i = 0; i < n_amaxtb; i++) {
      std::cout << "\t" << amaxtb_x[i] << "\t" << amaxtb_y[i] << std::endl;
    }

    // Send output to output channel
    // If there is an error, the flag will be negative
    flag = out_channel.send(1, &obj);
    if (flag < 0) {
      std::cout << "C++ Model: Error sending output." << std::endl;
      break;
    }

  }

  return 0;
}

